### PR TITLE
WIP: Use load immediate for address constants in Z JIT

### PR DESCRIPTION
**Pull request by Spencer-Comin**
Rewrite of `genLoadAddressConstant` to use load immediate instructions for address constants when not doing AOT.

Also defines `addMetaDataForCodeAddress` for RIL instructions to add instruction immediates containing addresses to PIC lists for patching on class unload and redefinition.

### CODE REVIEW

File: compiler/z/codegen/OMRTreeEvaluator.cpp [hunk 1]
```diff
@@ -2004,49 +2004,41 @@ genLoadAddressConstant(TR::CodeGenerator * cg, TR::Node * node, uintptr_t value,
...
-   if (cg->comp()->target().is64Bit())
+   if (comp->getOption(TR_EnableHCR))
       {
-      return genLoadLongConstant(cg, node, (uint64_t) value, targetRegister, cursor, cond, base);
+      comp->getStaticHCRPICSites()->push_front(cursor);
+      assumePatch = true;
       }
-   else
+
+   if (comp->target().is64Bit() && !isCompressedClassPointer && (assumePatch || over32Bits))
```
Comment (Spencer-Comin) [inline compiler/z/codegen/OMRTreeEvaluator.cpp:2035]: I'm assuming here that a class pointer in compressed refs mode will always fit in 32 bits. @0xdaryl, I've been told you are looking at changing that on Aarch64 MacOSX. Are those changes specific to that target, or will they affect my assumption here eventually?

Comment (r30shah) [reply compiler/z/codegen/OMRTreeEvaluator.cpp:2035]: Can we place a comment here for the taken assumption? Also I would wait for what @0xdaryl  says, but if that the assumption is that class pointer constants are 32 bit for compressed refs, least we should have an ASSERT for that.

File: compiler/z/codegen/OMRTreeEvaluator.cpp [hunk 2]
```diff
@@ -2004,49 +2004,41 @@ genLoadAddressConstant(TR::CodeGenerator * cg, TR::Node * node, uintptr_t value,
...
       {
-      // TODO: We should update this API as well to accept a "canSetConditionCode" and forward that argument here rather than passing false blindly.
-      return generateLoad32BitConstant(cg, node, (int32_t) value, targetRegister, false, cursor, cond, base);
+      int32_t high32 = value >> 32;
```
Comment (r30shah) [inline compiler/z/codegen/OMRTreeEvaluator.cpp:2037]: Can we static_cast the result of `value >> 32` to `uint32_t` , and also change the type of `high32` to `uint32_t`

File: compiler/z/codegen/S390Instruction.cpp [hunk 3]
```diff
@@ -1806,6 +1806,35 @@ TR::S390RILInstruction::adjustCallOffsetWithTrampoline(int32_t offset, uint8_t *
    return offsetHalfWords;
    }
 
+void
+TR::S390RILInstruction::addMetaDataForCodeAddress(uint8_t *cursor)
+   {
+   uintptr_t immediateAsAddress = (uintptr_t) (uint32_t) getSourceImmediate();
```
Comment (r30shah) [inline compiler/z/codegen/S390Instruction.cpp:1812]: We should use `static_cast` here to type cast return value of `getSourceImmediate()`. Also in most the function, we should avoid direct castings and use `static_cast`[1] or `reinterpret_cast`[2] where suitable. 

[1]. https://en.cppreference.com/w/cpp/language/static_cast
[2]. https://en.cppreference.com/w/cpp/language/reinterpret_cast

File: compiler/z/codegen/S390Instruction.cpp [hunk 4]
```diff
@@ -1806,6 +1806,35 @@ TR::S390RILInstruction::adjustCallOffsetWithTrampoline(int32_t offset, uint8_t *
...
+   if (getNext()->getOpCode().getOpCodeValue() == TR::InstOpCode::IIHF)
+      immediateAsAddress |= ((uintptr_t) toS390RILInstruction(getNext())->getSourceImmediate()) << 32;
+
+   TR::Compilation *comp = cg()->comp();
+   bool staticPIC = std::find(comp->getStaticPICSites()->begin(), comp->getStaticPICSites()->end(), this) != comp->getStaticPICSites()->end();
+   bool staticMethodPIC = std::find(comp->getStaticMethodPICSites()->begin(), comp->getStaticMethodPICSites()->end(), this) != comp->getStaticMethodPICSites()->end();
+   bool staticHCRPIC = std::find(comp->getStaticHCRPICSites()->begin(), comp->getStaticHCRPICSites()->end(), this) != comp->getStaticHCRPICSites()->end();
```
Comment (r30shah) [inline compiler/z/codegen/S390Instruction.cpp:1819]: We do not need to search for both staticPICSites and StaticMethodPICSites. If we have it in one, we can not have it in another right ??

File: compiler/z/codegen/S390Instruction.cpp [hunk 5]
```diff
@@ -1806,6 +1806,35 @@ TR::S390RILInstruction::adjustCallOffsetWithTrampoline(int32_t offset, uint8_t *
...
+
+   if (staticPIC && immediateAsAddress && !TR::Compiler->cls.sameClassLoaders(comp, (TR_OpaqueClassBlock *) immediateAsAddress, comp->getCurrentMethod()->classOfMethod()))
```
Comment (r30shah) [inline compiler/z/codegen/S390Instruction.cpp:1821]: In all three cases here, 32 Bit version of the Runtime helper to added location to be patched on class unload and redefinition. Do we handle it correctly? i.e, It seems to be creating the Unloading site and Redefinition site with size 4, I think `compensate` method in the RAT should take care of it but can you summarize how it works with 64-bit address?

**Review by r30shah — CHANGES_REQUESTED** [review]

File: compiler/z/codegen/S390Instruction.cpp (hunk 5 above)
Comment (Spencer-Comin) [reply compiler/z/codegen/S390Instruction.cpp:1821]: I've modified the `compensate` methods in https://github.com/eclipse-openj9/openj9/pull/15705 to have a special case for `LLILF`/`IIHF`. After patching the PIC location, it checks if it is in an `LLILF` and the next instruction is an `IIHF`. If it is, it also patches the `IIHF` immediate to hold the upper 32 bits of the new address (in HCR case) or -1 (in unload case).

File: compiler/z/codegen/OMRTreeEvaluator.cpp [hunk 6]
```diff
@@ -2004,49 +2004,41 @@ genLoadAddressConstant(TR::CodeGenerator * cg, TR::Node * node, uintptr_t value,
       return generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister, (uintptr_t) value, reloKind, cond, cursor, base);
       }
 
-   TR::Symbol *symbol = NULL;
-   TR::Compilation *comp = cg->comp();
+   if (node->getOpCode().hasSymbolReference())
```
[the remark this answers is not in the document]
Comment (Spencer-Comin) [reply compiler/z/codegen/OMRTreeEvaluator.cpp:2007]: Not currently, but if https://github.com/eclipse-openj9/openj9/pull/14932 goes through it will be

File: compiler/z/codegen/S390Instruction.cpp (hunk 5 above)
Comment (r30shah) [reply compiler/z/codegen/S390Instruction.cpp]: Okk, I think this is potential bug. I can not think of the example right away but we might have subsequent instruction after LLILF that is unrelated, and this change might update that as well. You should set the size parameter correctly when class and instruction address to list of PIC sites, which creates a runtime assumption. Also update the `compensate` in https://github.com/eclipse-openj9/openj9/pull/15705 to account for that.

Comment (Spencer-Comin) [reply compiler/z/codegen/S390Instruction.cpp]: OK I've reverted this PR to draft while I look into how to prevent this bug

Comment (Spencer-Comin) [reply compiler/z/codegen/S390Instruction.cpp]: Together with the update in https://github.com/eclipse-openj9/openj9/pull/15705, this bug should be averted now. The LLILF of the LLILF/IIHF pair is added as a 64 bit pic site (32 bit constants are still just in LLILF and added as 32 bit pic site), and the `compensate` will only update a subsequent IIHF when updating 64 bit pic sites.

**Comment by Spencer-Comin** [comment]
I'm assuming the MacOS failure is unrelated as this is a Z codegen change

**Comment by Spencer-Comin** [comment]
FYI @joransiu @r30shah

File: compiler/z/codegen/OMRTreeEvaluator.cpp [hunk 7]
```diff
@@ -2187,12 +2187,32 @@ OMR::Z::TreeEvaluator::lbitpermuteEvaluator(TR::Node *node, TR::CodeGenerator *c
    return TR::TreeEvaluator::bitpermuteEvaluator(node, cg);
    }
 
+static TR_ExternalRelocationTargetKind
+getRelocationTargetKindFromSymbol(TR::CodeGenerator* cg, TR::Symbol *sym)
```
Comment (r30shah) [inline compiler/z/codegen/OMRTreeEvaluator.cpp:2191]: @Spencer-Comin  I think this change may remove relocations being added by @cjjdespres in https://github.com/eclipse/omr/pull/6956/files#diff-322438948e52d98d89adb41b9fac54130800bf31d82634aa2e9bef9988400f20. Can you make sure that this routine also checks if the symbol is `CatchBlockCounter`  ?

**Review by r30shah — CHANGES_REQUESTED** [review]

File: compiler/z/codegen/S390Instruction.cpp [hunk 8]
```diff
@@ -2041,29 +2086,6 @@ TR::S390RILInstruction::generateBinaryEncoding()
       }
    else if (getOpCode().getOpCodeValue() == TR::InstOpCode::LARL)
       {
-      // Check if this LARL is loading unloadable constant
```
Comment (r30shah) [inline compiler/z/codegen/S390Instruction.cpp:2044]: Confirming, we would not be using `LARL` where it is used for static PIC sites or method sites at all, right?

**Review by r30shah — APPROVED** [review]
Went through the changes, LGTM, Can you verify if you have run sanity tests with these changes?

**Comment by 0xdaryl** [comment]
What are the next steps on this to get it merged?

**Comment by 0xdaryl** [comment]
@r30shah : For some reason I missed seeing your review updates from yesterday.  Sorry about that.  Glad you're moving things forward.

**Comment by Spencer-Comin** [comment]
> @Spencer-Comin can you mark this one and other OpenJ9 PR (one that uses changes made in this PR) as WIP till we get [eclipse-openj9/openj9#15705](https://github.com/eclipse-openj9/openj9/pull/15705) reviewed and merged?

FYI the OpenJ9 PR that relies on this change is https://github.com/eclipse-openj9/openj9/pull/14932

File: compiler/z/codegen/S390Instruction.cpp (hunk 8 above)
Comment (Spencer-Comin) [reply compiler/z/codegen/S390Instruction.cpp]: With https://github.com/eclipse-openj9/openj9/pull/14932 all uses of `LARL` for PIC sites will be removed, *except* for a usage in OMR (`aloadHelper()`) that I just found while double checking the codebase. I've amended my commit to leave this code in.

**Comment by Spencer-Comin** [comment]
Sanity tests have passed with these changes

**Comment by Spencer-Comin** [comment]
Sanity tests have passed with these changes

**Comment by r30shah** [comment]
@Spencer-Comin  Confirming, as https://github.com/eclipse-openj9/openj9/pull/15705 is merged, this one is good to go right?

**Comment by r30shah** [comment]
@Spencer-Comin can you rebase the change?  I will launch the sanity builds after that

**Comment by r30shah** [comment]
Jenkins build zos,zlinux

**Comment by r30shah** [comment]
@0xdaryl  This one is good to merge and Spencer has done all sanity testing. Can I request you to merge this PR ?
